oracle.go

```diff
@@ -33,280 +31,235 @@ DROP USER {{username}};
-	dbType, err := New()
+func Run() error {
+	db, err := New()
 	if err != nil {
 		return err
 	}
 
-	dbplugin.Serve(dbType.(dbplugin.Database), api.VaultPluginTLSProvider(apiTLSConfig))
+	newdbplugin.Serve(db.(newdbplugin.Database))
 
 	return nil
 }
 
-func (o *Oracle) Type() (string, error) {
-	return oracleTypeName, nil
+func (o *Oracle) Initialize(ctx context.Context, req newdbplugin.InitializeRequest) (newdbplugin.InitializeResponse, error) {
+	err := o.SQLConnectionProducer.Initialize(ctx, req.Config, req.VerifyConnection)
+	if err != nil {
+		return newdbplugin.InitializeResponse{}, err
+	}
+	resp := newdbplugin.InitializeResponse{
+		Config: req.Config,
+	}
+	return resp, nil
 }
 
-func (o *Oracle) CreateUser(ctx context.Context, statements dbplugin.Statements, usernameConfig dbplugin.UsernameConfig, expiration time.Time) (username string, password string, err error) {
-	statements = dbutil.StatementCompatibilityHelper(statements)
-
-	if len(statements.Creation) == 0 {
-		return "", "", dbutil.ErrEmptyCreationStatement
+func (o *Oracle) NewUser(ctx context.Context, req newdbplugin.NewUserRequest) (newdbplugin.NewUserResponse, error) {
+	statements := removeEmpty(req.Statements.Commands)
+	if len(statements) == 0 {
+		return newdbplugin.NewUserResponse{}, dbutil.ErrEmptyCreationStatement
 	}
 
-	// Grab the lock
 	o.Lock()
 	defer o.Unlock()
 
-	username, err = o.GenerateUsername(usernameConfig)
+	username, err := newUsername(req.UsernameConfig)
 	if err != nil {
-		return "", "", err
+		return newdbplugin.NewUserResponse{}, fmt.Errorf("failed to generate username: %w", err)
 	}
 	username = strings.ToUpper(username)
 
-	password, err = o.GeneratePassword()
+	db, err := o.getConnection(ctx)
 	if err != nil {
-		return "", "", err
+		return newdbplugin.NewUserResponse{}, fmt.Errorf("failed to get connection: %w", err)
 	}
 
-	expirationStr, err := o.GenerateExpiration(expiration)
+	err = newUser(ctx, db, username, req.Password, req.Expiration, req.Statements.Commands)
 	if err != nil {
-		return "", "", err
+		return newdbplugin.NewUserResponse{}, err
 	}
 
-	// Get the connection
-	db, err := o.getConnection(ctx)
-	if err != nil {
-		return "", "", err
+	resp := newdbplugin.NewUserResponse{
+		Username: username,
 	}
+	return resp, nil
+}
 
-	// Start a transaction
-	tx, err := db.Begin()
-	if err != nil {
-		return "", "", err
-
+func removeEmpty(strs []string) []string {
+	newStrs := []string{}
+	for _, str := range strs {
+		str = strings.TrimSpace(str)
+		if str == "" {
+			continue
+		}
+		newStrs = append(newStrs, str)
 	}
-	// Effectively a no-op if the transaction commits successfully
-	defer tx.Rollback()
+	return newStrs
+}
 
-	// Execute each query
-	for _, stmt := range statements.Creation {
-		for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
-			query = strings.TrimSpace(query)
-			if len(query) == 0 {
-				continue
-			}
+func newUsername(config newdbplugin.UsernameMetadata) (string, error) {
```

Comment:
I'm wondering if `newUsername()` and associated functions should live in dbutil somewhere? IIRC the converted mongodb plugin is using something similar.

Reply:
After talking a bit offline, we decided to move the existing `SQLCredentialsProducer` username generation logic into its own function that can be used here: https://github.com/hashicorp/vault/pull/10050